### ISSUE TITLE
Address of memory circuits is limited to 24 bits

### ISSUE THREAD

**Issue by sebastianardelean** [hneemann/Digital#953]
After using the memory circuits (RAM/EEPROM) I observed that even if the option to have 32 bits address is available, the maximum accepted value is 24

**Comment by hneemann** [comment]
There must be a limit, otherwise users would enter 64 bits and wonder why there are problems with the simulation. :-)
And 16MBytes of RAM seemed to be enough for a simulation.

**Comment by sebastianardelean** [comment]
@hneemann  
Hi,

That's ok and is a good reason. But Digital allows setting the address on 32 bits (the options in the dropdown menu are: 8,16,32). But if you chose the 32bits option, after pressing save, the address defaults to 24, a value that is not in the list. I agree with you, from the perspective of simulation. But the question is: what if I want to implement a 32-bit processor and generate the Verilog/VHDL code. In that case, I'm not interested in simulation, but in the generation of a valid code using valid components.

My suggestion would be: for the cases where the address is bigger than 16bits, the simulation option shouldn't be available. But I would have the option to create circuits with 32 bits buses and generate the code for them. 

Best Regards,
Sebastian

**Comment by hneemann** [comment]
@sebastianardelean 
This would make little sense for most memory components, since these can only be synthesized as "distrubuted ram". Only for the "Block RAM" component can actually also block ram be used in the FPGA. For this component, it might be reasonable to allow more address bits. I have to think about it.

Secondly, the simulator is not really a design tool. It is intended for teaching purposes. The focus is rather to be able to run all built circuits on an FPGA. Not so much to actually be able to use all the features of an FPGA.

**Comment by prlaba** [comment]
IMHO the “bug” here is having the value 32 appear as the max value in the drop-down list when in fact it isn’t (24 is).

That’s not good human factors design.  If the maximum value is 24 (for whatever reason), show *that* value as the maximum in the drop-down list, not some value that can’t actually be selected.

Why not simply include *all* possible values in the drop-down list, from min to max?  That makes clear what values are permitted. (It was confusing to me at first that I could still assign values between 16 and 32 (24), even though they weren’t listed.)  In general, drop-down lists should include all possible values, in some meaningful order (in this case lowest-to highest bits).

I agree with your comment that 24 address bits is large enough for Digital’s purposes.  But I know of users out there building CPU’s or full systems in simulators only that might want to simulate larger RAM.  Just out of curiosity, what’s the “cost” of supporting more than 24 bits?  Performance?  Memory?  Both?

Thanks!

**Comment by hneemann** [comment]
> IMHO the “bug” here is having the value 32 appear as the max value in the drop-down list when in fact it isn’t (24 is).

This issue is already fixed.

> Why not simply include all possible values in the drop-down list, from min to max?

This is not always possible without the list becoming too long.
Therefore, only the frequently used values are selectable from the list, others must be entered manually.

> Just out of curiosity, what’s the “cost” of supporting more than 24 bits? Performance? Memory? Both?

It is only a question of memory consumption. Regarding the perfomance, it does not matter how large the memory is. The only thing to keep in mind is that internally always 64-bit values are stored. If 32 address bits were used, the memory block allocated in the heap would already have a size of 32 GBytes.

**Comment by sebastianardelean** [comment]
@hneemann Thank you for the clarifications! I like this tool, right now I'm using for designing circuits, as a hobby, but I plan to use it for teaching. Thank you!

**Comment by prlaba** [comment]
Thanks Helmut,

 

Re your lists becoming too long: Drop-down lists are scrollable, so I don’t see the issue by including all possible values.  Which of Digital’s drop-down lists would become exceedingly long?

 

Soapbox on…

 

Again, having a dop-down list that shows some but not all of its selectable values (or showing values that are not assigned) is not good human factors design.  A user would (incorrectly) assume that the only possible values are those shown in the list.  (I made that mistaken assumption myself the first time I tried to create an EEPROM components with 19 address bits.  It was only after playing around a bit that I realized I could ignore the drop-down box and enter the number of bits directly in the text box.)

 

Soapbox off…

 

Yes, I figured each bit increase in a memory component’s address bits would double its size in memory.  I simulate a dozen or so fairly large EEPROMS (256K x 8 and 512K x 8) and they do consume a lot of memory.

 

Thanks,

 

Paul
